### PR TITLE
Improve Autofill UI and enable inline presentations

### PR DESCRIPTION
**Pull request by fmeum**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## :scroll: Description

Improves the Autofill UI in the following ways:

* Add support for Android 11 inline presentations of Autofill datasets.
* Instead of showing the identifier (app name or web origin) of the
current app on top of every Autofill dataset, it is now shown 1) as a
header dataset on Android 9 and 10 as well as 2) at the top of the
search activity on all supported versions of Android. Rationale: The
identifier is only used in trust decisions when choosing an existing
entry to fill and should feature prominently in that view, not
elsewhere.
* Show the actual identifier part of a matched entry's path, which may
differ from the identifier of the matched app/website.
* Slightly tweak the labels of Search/Generate Autofill actions to
indicate that a) this is about entries and b) the user may skip the
generation of a password and supply a custom one as well.

## :green_heart: How did you test it?
Went through the typical use cases and confirmed that they still work as intended.
When filling an entry and then clearing the form manually, subsequent Autofill UI will show the usual actions minus the first one (this is repeatable). I am investigating.

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [ ] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

## :crystal_ball: Next steps

- [ ] Decide whether we want an option to disable inline presentations
- [ ] Come up with good changelog entries
- [ ] Fix the bug described above

### CODE REVIEW

**Comment by fmeum** [comment]
Forgot to include screenshots:
Screenshot_20201102-121419
Screenshot_20201102-121301
Screenshot_20201102-112741

**Comment by fmeum** [comment]
> When filling an entry and then clearing the form manually, subsequent Autofill UI will show the usual actions minus the first one (this is repeatable). I am investigating.

This plays out without invocations of our service in between. May very well be an AOSP Autofill bug, as if we needed more of those...

File: app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt [hunk 1]
```diff
@@ -162,6 +162,12 @@ class AutofillFilterView : AppCompatActivity() {
                 setText(initialSearch, TextView.BufferType.EDITABLE)
                 addTextChangedListener { updateSearch() }
             }
+            origin.text = buildSpannedString {
+                append("Search and fill into\n")
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt]: Expose for localisation?

File: app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/InlineResponseBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.zeapo.pwdstore.autofill.oreo.ui
+
+import android.app.PendingIntent
+import android.graphics.drawable.Icon
+import android.os.Build
+import android.service.autofill.InlinePresentation
+import android.widget.inline.InlinePresentationSpec
+import androidx.autofill.inline.UiVersions
+import androidx.autofill.inline.v1.InlineSuggestionUi
+
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/InlineResponseBuilder.kt]: I guess this file shouldn't have been here

Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/InlineResponseBuilder.kt]: Right, deleted it.

File: app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt (hunk 1 above)
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt]: Should be done, added a German translation for good measure

**Review by msfjarvis — DISMISSED** [review]
LGTM, will merge as soon as I can test this.

**Comment by fmeum** [comment]
> > When filling an entry and then clearing the form manually, subsequent Autofill UI will show the usual actions minus the first one (this is repeatable). I am investigating.
> 
> This plays out without invocations of our service in between. May very well be an AOSP Autofill bug, as if we needed more of those...

This behavior is documented [here](https://developer.android.com/reference/android/service/autofill/Dataset.Builder#setAuthentication(android.content.IntentSender)). Our fill-in-only dataset replaces the original dataset, but as it doesn't come with a visual representation, it is no longer shown in the fill UI. This could be fixed by returning the original dataset + filled in data, but that will need to be passed around for quite a bit. @msfjarvis Do we care?

**Comment by msfjarvis** [comment]
> > > When filling an entry and then clearing the form manually, subsequent Autofill UI will show the usual actions minus the first one (this is repeatable). I am investigating.
> > 
> > This plays out without invocations of our service in between. May very well be an AOSP Autofill bug, as if we needed more of those...
> 
> This behavior is documented [here](https://developer.android.com/reference/android/service/autofill/Dataset.Builder#setAuthentication(android.content.IntentSender)). Our fill-in-only dataset replaces the original dataset, but as it doesn't come with a visual representation, it is no longer shown in the fill UI. This could be fixed by returning the original dataset + filled in data, but that will need to be passed around for quite a bit. @msfjarvis Do we care?

I don't think so, will let you know if I feel differently when putting this to use.

**Comment by fmeum** [comment]
I added a fixme comment for posterity. It's surprisingly cumbersome to clone and modify a dataset, I fear this would require us adding our own `Dataset.Builder` clone with a `Parcelable` implementation.

**Review by msfjarvis — APPROVED** [review]
Works very well :ok_hand:

**Comment by msfjarvis** [comment]
Hm I didn't test it with apps where it seems to be crashing, failing with `java.lang.IllegalStateException: at least one value must be set`.

Edit: I can repro on Twitter's automatically detected login form and by manually triggering Autofill in Instagram's login page.

Edit 2: unproguarded stacktrace

<details><summary>Stacktrace</summary>

```
 AndroidRuntime  D  Shutting down VM
                 E  FATAL EXCEPTION: main
                 E  Process: dev.msfjarvis.aps, PID: 29196
                 E  java.lang.IllegalStateException: at least one value must be set
                 E      at android.service.autofill.Dataset$Builder.build(Dataset.java:663)
                 E      at com.zeapo.pwdstore.autofill.oreo.AutofillResponseBuilder.makeIntentDataset(AutofillResponseBuilder.kt:60)
                 E      at com.zeapo.pwdstore.autofill.oreo.AutofillResponseBuilder.makeFillOtpFromSmsDataset(AutofillResponseBuilder.kt:91)
                 E      at com.zeapo.pwdstore.autofill.oreo.AutofillResponseBuilder.makeFillResponse(AutofillResponseBuilder.kt:167)
                 E      at com.zeapo.pwdstore.autofill.oreo.AutofillResponseBuilder.fillCredentials(AutofillResponseBuilder.kt:188)
                 E      at com.zeapo.pwdstore.autofill.oreo.OreoAutofillService.onFillRequest(OreoAutofillService.kt:95)
                 E      at android.service.autofill.-$$Lambda$I0gCKFrBTO70VZfSZTq2fj-wyG8.accept(Unknown Source:8)
                 E      at com.android.internal.util.function.pooled.PooledLambdaImpl.doInvoke(PooledLambdaImpl.java:307)
                 E      at com.android.internal.util.function.pooled.PooledLambdaImpl.invoke(PooledLambdaImpl.java:201)
                 E      at com.android.internal.util.function.pooled.OmniFunction.run(OmniFunction.java:97)
                 E      at android.os.Handler.handleCallback(Handler.java:938)
                 E      at android.os.Handler.dispatchMessage(Handler.java:99)
                 E      at android.os.Looper.loop(Looper.java:223)
                 E      at android.app.ActivityThread.main(ActivityThread.java:7656)
                 E      at java.lang.reflect.Method.invoke(Native Method)
                 E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
                 E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

</details>

**Comment by fmeum** [comment]
> Hm I didn't test it with apps where it seems to be crashing, failing with `java.lang.IllegalStateException: at least one value must be set`.
> 
> Edit: I can repro on Twitter's automatically detected login form and by manually triggering Autofill in Instagram's login page.
> 
> Edit 2: unproguarded stacktrace
> 
> Stacktrace

Could you try disabling and re-enabling Password Store's Autofill feature and then record a video for me? I am unable to reproduce this in either combination of web/native app and Twitter/Instagram. In fact I tested the new functionality with Twitter's native app from the very beginning. 

The stacktrace does scare me as it seems we are creating Datasets without any fields set, which the code should explicity check for and prevent...

**Comment by msfjarvis** [comment]
> > Hm I didn't test it with apps where it seems to be crashing, failing with `java.lang.IllegalStateException: at least one value must be set`.
> > Edit: I can repro on Twitter's automatically detected login form and by manually triggering Autofill in Instagram's login page.
> > Edit 2: unproguarded stacktrace
> > Stacktrace
> 
> Could you try disabling and re-enabling Password Store's Autofill feature and then record a video for me? I am unable to reproduce this in either combination of web/native app and Twitter/Instagram. In fact I tested the new functionality with Twitter's native app from the very beginning.
> 
> The stacktrace does scare me as it seems we are creating Datasets without any fields set, which the code should explicity check for and prevent...

Toggling it off and back on fixed the issue, but we can't make all users do the same :(

**Comment by fmeum** [comment]
Given that we didn't change anything about the `Dataset`s we generate other than their visual representation, I don't see how we can be responsible for this issue. It may also just be an issue with debug builds. If we can reproduce this issue with actual release APKs, we should provide an onboarding flow that disables and then asks the user to re-enable Autofill.

**Comment by msfjarvis** [comment]
> Given that we didn't change anything about the `Dataset`s we generate other than their visual representation, I don't see how we can be responsible for this issue. It may also just be an issue with debug builds. If we can reproduce this issue with actual release APKs, we should provide an onboarding flow that disables and then asks the user to re-enable Autofill.

Sounds good to me.

**Comment by fmeum** [comment]
I gave this a try with the following APKs: https://drive.google.com/file/d/1mJIyu5NNIPtRbrwPEH82-RTOalOfPbF1/view?usp=sharing, https://drive.google.com/file/d/1o0oOveur9gZX8JnJNvjpL-E25U_NSX5w/view?usp=sharing
I wasn't able to reproduce the issue, @msfjarvis let me know whether you could.

**Comment by msfjarvis** [comment]
> I gave this a try with the following APKs: [drive.google.com/file/d/1mJIyu5NNIPtRbrwPEH82-RTOalOfPbF1/view?usp=sharing](https://drive.google.com/file/d/1mJIyu5NNIPtRbrwPEH82-RTOalOfPbF1/view?usp=sharing), [drive.google.com/file/d/1o0oOveur9gZX8JnJNvjpL-E25U_NSX5w/view?usp=sharing](https://drive.google.com/file/d/1o0oOveur9gZX8JnJNvjpL-E25U_NSX5w/view?usp=sharing)
> I wasn't able to reproduce the issue, @msfjarvis let me know whether you could.

I can't repro on debug builds, but can consistently do so on release ones. Autofill crashed with the same stacktrace when I was setting up your debug build w/ password auth, after I had upgraded the release build I use on my device to the tip of `develop`.

**Comment by fmeum** [comment]
Could you get the log output before the crash for me? This should show us which `AutofillScenario` is detected and whether there is any logic bug that results in `Dataset`s not being set up correctly.

**Comment by msfjarvis** [comment]
> Could you get the log output before the crash for me? This should show us which `AutofillScenario` is detected and whether there is any logic bug that results in `Dataset`s not being set up correctly.

[Here you go](https://bin.kv2.dev/~5fa119a1b441021526f2293c?raw=true)

**Pull request merged**